chaincode/compute_plan.go

```diff
@@ -0,0 +1,251 @@
+			out := outputTraintuple{}
+			err = out.Fill(db, traintuple, traintupleKey)
+			if err != nil {
+				return resp, err
+			}
+			traintuplesTodo = append(traintuplesTodo, out)
+		}
+
+		err = traintuple.Save(db, traintupleKey)
+		if err != nil {
+			return resp, err
+		}
+		traintupleKeysByID[computeTraintuple.ID] = traintupleKey
+		resp.TraintupleKeys = append(resp.TraintupleKeys, traintupleKey)
+	}
+
+	resp.TesttupleKeys = []string{}
+	for index, computeTesttuple := range inp.Testtuples {
+		traintupleKey, ok := traintupleKeysByID[computeTesttuple.TraintupleID]
+		if !ok {
+			return resp, errors.BadRequest("testtuple index %s: traintuple ID %s not found", index, computeTesttuple.TraintupleID)
+		}
+		testtuple := Testtuple{}
+		testtuple.Model = &Model{TraintupleKey: traintupleKey}
+		testtuple.ObjectiveKey = inp.ObjectiveKey
+		testtuple.AlgoKey = inp.AlgoKey
+
+		inputTesttuple := inputTesttuple{}
+		inputTesttuple.DataManagerKey = computeTesttuple.DataManagerKey
+		inputTesttuple.DataSampleKeys = computeTesttuple.DataSampleKeys
+		inputTesttuple.Tag = computeTesttuple.Tag
+		err = testtuple.SetFromInput(db, inputTesttuple)
+		if err != nil {
+			return resp, err
+		}
+		testtuple.Status = StatusWaiting
+		testtupleKey := testtuple.GetKey()
+		err = testtuple.Save(db, testtupleKey)
+		if err != nil {
+			return resp, err
+		}
+		resp.TesttupleKeys = append(resp.TesttupleKeys, testtupleKey)
+	}
+
+	event := TuplesEvent{}
+	event.SetTraintuples(traintuplesTodo...)
+	err = SendTuplesEvent(db.cc, event)
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, err
+}
+
+func queryComputePlan(db LedgerDB, args []string) (resp outputComputePlanDetails, err error) {
+	inp := inputHash{}
+	err = AssetFromJSON(args, &inp)
+	if err != nil {
+		return
+	}
+	return getComputePlan(db, inp.Key)
+}
+
+func queryComputePlans(db LedgerDB, args []string) (resp []outputComputePlanDetails, err error) {
+	computePlanKeys, err := db.GetIndexKeys("traintuple~computeplanid", []string{"computeplan"})
+	if err != nil {
+		return
+	}
+	for _, key := range computePlanKeys {
+		var computePlan outputComputePlanDetails
+		computePlan, err = getComputePlan(db, key)
+		if err != nil {
+			return
+		}
+		resp = append(resp, computePlan)
+	}
+	return resp, err
+}
+
+func getComputePlan(db LedgerDB, key string) (resp outputComputePlanDetails, err error) {
+	// get traintuples
+	traintuples, err := getTraintuplesForComputePlan(db, key)
+	if err != nil {
+		return
+	}
+	if len(traintuples) == 0 {
+		err = errors.E("No traintuple found for compute plan %s", key)
+		return
+	}
+	var firstTt *Traintuple
+	ttKeys := []string{}
+	outTts := []outputTraintuple{}
+	for _, traintupleAndKey := range traintuples {
+		ttKey := traintupleAndKey.Key
+		tt := traintupleAndKey.Traintuple
+
+		if firstTt == nil {
+			firstTt = &tt
+		}
+		ttKeys = append(ttKeys, ttKey)
+
+		outTt := outputTraintuple{}
+		outTt.Fill(db, tt, ttKey)
+		outTts = append(outTts, outTt)
+	}
```

Comment:
I have the impression we range over that list twice and it could be avoided if `getTraintupleForComputePlan` would return `OutputTraintuple` already filled. But I could clearly be missing something.

Reply:
Now that we return keys instead of `outputTesttuple`s, there's no double loop anymore :)